### DISCUSSION TITLE
Performance of MCA backends: MPFR faster than QUAD

### DISCUSSION THREAD

**Discussion by yohanchatelain** [verificarlo/verificarlo#284] [Q&A] [answered]
I compared the performance of MCA backends MPFR and QUAD and it seems that MPFR is faster than QUAD while I expected the oposite.
Here are the results of the comparison for the `test_fortran_NAS`, `time ./test.sh`:
`libinterflop_ieee.so` 1.7s
`libinterflop_mca.so` 23.7s
`libinterflop_mca_mpfr.so` 18.5s
So MPFR seems to be 30% faster.
I attached the [test.gz](https://github.com/verificarlo/verificarlo/files/5403185/test.gz)

-- config --
Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz

**Comment by pablooliveira** [comment]
Here are the timings in my system  Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz with 8Gb ram (I repeated runs two times) :

libinterflop_mca.so 26.55, 26.53
libinterflop_mca_mpfr.so 23.93, 23.23 

So indeed it seems mca is slower on this benchmark than mca_mpfr.
This is not critical but we should benchmark both implementations again.

I remember having benchmarked on my same exact system both backends and getting faster results with mca quad. Therefore it is not linked to Yohan's system. 

As discussed in our last meeting, some hypothesis for this could be:
- performance regression in mca (quad)
- performance improvement in latest version of mpfr

**Comment by pablooliveira** [comment]
And thanks for packaging the benchmark @yohanchatelain !

**Comment by mistoan** [comment]
@yohanchatelain @pablooliveira I'm re-opening this discussion on the performance of the MCA and MCA-MPFR backends.

The context is the change in the random number generator.
All backends that use the RNG now use the `drand48_r` (and `lrand48_r` where necessary).
The generator is re-entrant.

The test consists of performing the sum of 1e6 randomly generated numbers, and the sum is repeated 1e3 times.
The threads divide the work equally between themselves.
The threads are generated using OpenMP.
There are two versions for the test, one where the sum is naive, the other uses the Kahan sum.
I've included the tests here [test_parallel_sum.tar.gz](https://github.com/verificarlo/verificarlo/files/7338541/test_parallel_sum.tar.gz).

The behavior that I observe is that there is a 5x factor between MCA and MCA-MPFR.
This factor seems to remain approximately the same as the number of threads increases.

Here are some results on my machine, a system based around an Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz x 6
```
Naive Sum - OpenMP
- float
    - 1e3 * 1e6 elements

        - original: 17.952s
        - vfc ieee: 19.003s
        - vfc mca:  45.184s

        - 2 threads
        - original: 9.118s
        - vfc ieee: 10.009s
        - vfc mca:  23.912s

        - 4 threads
        - original: 5.112s
        - vfc ieee: 6.033s
        - vfc mca:  14.207s

        - 6 threads
        - original: 4.032s
        - vfc ieee: 4.858s
        - vfc mca:  11.137s

        - 8 threads
        - original: 4.262s
        - vfc ieee: 5.249s
        - vfc mca:  11.500s

        - 12 threads
        - original: 5.348s
        - vfc ieee: 6.518s
        - vfc mca:  11.925
```

```
Kahan Sum - OpenMP
    - float
    - 1e3 * 1e6 elements

        - original: 20.311s
        - vfc ieee: 33.045s
        - vfc mca:  87.909s
        - vfc mca_mpfr: 427.347s

        - 2 threads
        - original: 10.518s
        - vfc ieee: 17.927s
        - vfc mca:  46.445s
        - vfc mca_mpfr: 229.125s

        - 4 threads
        - original: 5.823s
        - vfc ieee: 10.753s
        - vfc mca:  27.700s
        - vfc mca_mpfr: 133.875s

        - 6 threads
        - original: 4.384s
        - vfc ieee: 8.870s
        - vfc mca:  21.190s
        - vfc mca_mpfr: 101.160s

        - 8 threads
        - original: 4.784s
        - vfc ieee: 8.823s
        - vfc mca:  24.212s
        - vfc mca_mpfr: 128.914s

        - 12 threads
        - original: 5.290s
        - vfc ieee: 10.369s
        - vfc mca:  26.379s
        - vfc mca_mpfr: 119.658s
```

```
Kahan Sum - OpenMP
    - double
    - 1e3 * 1e6 elements

        - original: 22.025s
        - vfc ieee: 32.199s
        - vfc mca:  294.604s
        - vfc mca_mpfr: 432.448s

        - 2 threads
        - original: 11.307s
        - vfc ieee: 17.301s
        - vfc mca:  156.482s
        - vfc mca_mpfr: 229.997s

        - 4 threads
        - original: 6.002s
        - vfc ieee: 9.866s
        - vfc mca:  91.587s
        - vfc mca_mpfr: 138.731s

        - 6 threads
        - original: 4.606s
        - vfc ieee: 8.331s
        - vfc mca:  73.399s
        - vfc mca_mpfr: 122.730s

        - 8 threads
        - original: 4.511s
        - vfc ieee: 8.292s
        - vfc mca:  86.730s
        - vfc mca_mpfr: 139.130s

        - 12 threads
        - original: 5.020s
        - vfc ieee: 7.886s
        - vfc mca:  80.306s
        - vfc mca_mpfr: 132.259s
```

I would have said that some of the slowdown is due to how MPFR is handled in multi-threaded executions, but you can see that the slowdown is roughly the same even in the single-threaded execution.

**Reply by mistoan** [reply]
I've updated the tests with experiments on `double`.
There's a factor of roughly 3.5x between the executions in `float` and in `double`.
This is due to the fact that the simulations for doubles are done in `quad`.
However, even so the MCA backend has performance approximately 1.5x better then the MCA-MPFR backend.

**Comment by pablooliveira** [comment, marked as answer]
The new mca integer backend is now much faster and was recently further optimized by @yohanchatelain in #308. Moreover, the legacy mpfr backend was removed when moving to LLVM licence.
